pep-0577.rst

```diff
@@ -0,0 +1,652 @@
+outer scope before binding it locally.
+
+At function scope, augmented assignments will be changed to require that there
+be either a preceding name binding or variable declaration to explicitly
+establish the target name as being local to the function, or else an explicit
+``global`` or ``nonlocal`` declaration. ``TargetNameError``, a new
+``SyntaxError`` subclass, will be raised at compile time if no such binding or
+declaration is present.
+
+For example, the following code would compile and run as it does today::
+
+    x = 0
+    x += 1 # Sets global "x" to 1
+
+    class C:
+        x += 1 # Sets local "x" to 2, leaves global "x" alone
+
+    def local_target():
+        x = 0
+        x += 1 # Sets local "x" to 1, leaves global "x" alone
+
+    def global_target():
+        global x
+        x += 1 # Increments global "x" each time this runs
+
+    def nonlocal_target():
+        x = 0
+        def g():
+            nonlocal x
+            x += 1 # Increments "x" in outer scope each time this runs
+            return x
+        return g
+
+The follow examples would all still compile and then raise an error at runtime
+as they do today::
+
+    n += 1 # Raises NameError at runtime
+
+    class C:
+        n += 1 # Raises NameError at runtime
+
+    def missing_global():
+        global n
+        n += 1 # Raises NameError at runtime
+
+    def delayed_nonlocal_initialisation():
+        def f():
+            nonlocal n
+            n += 1
+        f() # Raises NameError at runtime
+        n = 0
+
+    def skipped_conditional_initialisation():
+        if False:
+            n = 0
+        n += 1 # Raises UnboundLocalError at runtime
+
+    def local_declaration_without_initial_assignment():
+        n : typing.Any
+        n += 1 # Raises UnboundLocalError at runtime
+
+Whereas the following would raise a compile time ``DeprecationWarning``
+initially, and eventually change to report a compile time ``TargetNameError``::
+
+    def missing_target():
+        x += 1 # Compile time TargetNameError due to ambiguous target scope
+               # Is there a missing initialisation of "x" here? Or a missing
+               # global or nonlocal declaration?
+
+As a conservative implementation approach, the compile time function name
+resolution change would be introduced as a ``DeprecationWarning`` in Python
+3.8, and then converted to ``TargetNameError`` in Python 3.9. This avoids
+potential problems in cases where an unused function would currently raise
+``UnboundLocalError`` if it was ever actually called, but the code is actually
+unused - converting that latent runtime defect to a compile time error qualifies
+as a backwards incompatible change that requires a deprecation period.
+
+When augmented assignments are used as expressions in function scope (rather
+than as standalone statements), there aren't any backwards compatibility
+concerns, so the compile time name binding checks would be enforced immediately
+in Python 3.9.
```

Comment:
Probably you meant 3.8 here?

It's important to address the syntactic ambiguity though -- a toplevel augmented assignment would be both an expression and a statement. (Possibly you already did and I've simply forgotten I'd already read your solution -- I was interrupted by much of a workday during this review.)

Reply:
I was being overly casual with terminology and saying "statement" when I really meant "top level expression in an expression statement". I'll clarify that.